---
title: "NanoporeCrunch"
output: html_document
---

```{r Libraries}
library(grid)
library(gtable)
library(gridExtra)
library(reshape2)
library(scales)
library(rlist)
library(readxl)
library(magrittr)
library(plyr)
library(dplyr)
library(ggplot2)
```


```{r Load csvs, xls}
# Load data, extract metadata from names
datasets <- list.files(path = ".", pattern = ".csv")
splits <- unlist(strsplit(datasets, "_"))

# Load supplementary files
expList <- read_excel("Experiment_List.xlsx")

  # Uses hg19  Not all cut sites matter for all runs
cutSites <- read_xlsx("Guides.xlsx")
cutSites <- data.frame(cutSites[,1:3], stringsAsFactors = F)
rownames(cutSites) <- NULL
cutSites$hg19 <- as.numeric(cutSites$hg19)
window = 50

rois <- read_xlsx("Regions_of_Interest.xlsx")


# Assign the easy things
dates <- grep("^[0-9]+$", splits, value = TRUE)
```

```{r Load experiment list}

# Next up: read these csvs and append metadata as new column
allData <- do.call(rbind, lapply(1:length(datasets), function(x) {
      working <- data.frame(read.csv(datasets[x], sep = ",", header = F, stringsAsFactors = F),
                            stringsAsFactors = F)
      working$Date <- dates[x]
      return(working)
      }
    ))

allData <- merge(allData, expList[, 1:3], by = "Date")

colnames(allData)[1:7] <- c("Date", "Chr", "Chr_S", "Chr_E", "Name", "Score", "Strand")

# Remove weird special characters that are somehow getting prepended onto each data set
allData$Chr <- gsub("[^a-zA-Z0-9]", "", allData$Chr)

# Chromosome lengths for plotting purposes.  Last three X, Y, M
chrlims <- c(247250000, 242952000, 199502000, 191274000, 180858000,
             170900000, 158822000, 146275000, 140274000, 135375000,
             134453000, 132350000, 114143000, 106369000, 100339000,
             88828000, 78775000, 76118000, 63812000, 62436000,
             46945000, 49692000, 154914000, 57773000, 16569)

# Length will be important later
  # Semi-arbitrarily filter at 50
allData <- allData %>% mutate(Len = Chr_E - Chr_S) %>% filter(Score >= 50)
```


```{r Crunch it!}

# Create columns for *all* cut sites
    # Later we can prune by which guides were present in which experiment

# Search Chr_S and Chr_E for proximity to each cut site

# For each guide
  # Filter dataUnion by Chr
  # Filter to dates in which that guide was used

cutDists <- lapply(1:nrow(cutSites), function(x) {
  working <- allData %>% filter(Chr == cutSites$Chr[x])
  dateList <- expList$Date[!(is.na(expList[, cutSites$Name[x]]))]
  working2 <- filter(working, as.numeric(Date) %in% dateList)
  working2[, paste0(cutSites$Name[x],"S")] <- c(abs(working2$Chr_S - cutSites$hg19[x]))
  working2[, paste0(cutSites$Name[x],"E")] <- c(abs(working2$Chr_E - cutSites$hg19[x]))
  working2
  }
)

allDists <- join_all(cutDists, type = "full")

allDists$rowMin <- sapply(1:nrow(allDists), function(x)
              min(allDists[x, 11:ncol(allDists)], na.rm = T))

# Filter by distance from cut site

withCuts <- filter(allDists, rowMin < 50)

withCuts$cutGuide <- sapply(1:nrow(withCuts), function(x)
        colnames(withCuts[x, 11:76])[match(withCuts$rowMin[x], withCuts[x, 11:76])])


whichCuts <- withCuts %>%
  select(c("Date", "Chr", "Chr_S", "Chr_E", "Name",
           "Score", "Strand","Sample", "Device", "Len", "cutGuide"))

# Collapse by start or end first
test <- whichCuts
test <- test %>% mutate(End = substr(cutGuide, nchar(cutGuide), nchar(cutGuide))) %>%
  mutate(cutGuide = substr(cutGuide, 0, nchar(cutGuide)-1))


test2 <- test %>% group_by(Date, cutGuide, Device) %>% dplyr::summarise(n())
test3 <- test %>% group_by(Date, Device) %>% dplyr::summarise(n())

```


```{r Check cuts from other guides on regions of interest}
# There's probably a way of doing all this stuff with `join`
  # See near the bottom of this chunk...

# Region of interest dataframe, then iterate over guides to see when they're cutting there

inGenes <- bind_rows(lapply(1:nrow(rois), function(x) {
            working <- allData %>% filter(Chr == rois$Chr[x])
            
            # Start of read in roi
            test <- intersect(which(working$Chr_S >= rois$Start[x]),
                            which(working$Chr_S <= rois$End[x]))
            # End of read in roi
            test2 <- intersect(which(working$Chr_E >= rois$Start[x]),
                            which(working$Chr_E <= rois$End[x]))
            # Start and end span roi (more rare)
            test3 <- intersect(which(working$Chr_S <= rois$Start[x]),
                            which(working$Chr_E >= rois$End[x]))
            
            # Duh.  Saves a lot of painful matching later
            working$Gene <- rois$Gene[x]
            
            # Less warning-prone than `union` with empty sets
            return(working[unique(c(test, test2, test3)), ])
      }
  ))


expGenes <- c(sapply(1:nrow(expList), function(x) {
  allGenes <- colnames(expList)[!is.na(expList[x, ])]
  allGenes <- allGenes[5:length(allGenes)]
  # Gene names are annoying to regex out of the excel file...
  junction <- ifelse(regexpr("([0-9][.,_])|([a-zA-Z][.,_])", allGenes) > 0,
                  regexpr("([0-9][.,_])|([a-zA-Z][.,_])", allGenes),
                  regexpr("[0-9][a-zA-Z]|([0-9][a-zA-Z])", allGenes))
  test2 <- substr(allGenes, 0,
                  ifelse(junction > 0, junction, nchar(allGenes)))
  rois$Gene[match(unique(test2),rois$Gene)]
}))

expGenes <- data.frame(cbind(unlist(expGenes),
                             rep(expList$Date, times = lengths(expGenes))),
                       stringsAsFactors = F)
colnames(expGenes) <- c("Gene", "Date")

# inGenes has reads in rois.  expGenes lists gene targets by date
  # anti_join removes combos of expGenes from inGenes.  Ta-da!
offTarget <- anti_join(inGenes, expGenes, by = c("Date", "Gene"))

# Example plot: find off-target reads from guides ordered in Nov
# PML = filter(offTarget, Gene == "PML")
# PML$num = 1:nrow(PML)
# 
# RARA = filter(offTarget, Gene == "RARA")
# RARA$num = 1:nrow(RARA)
# 
# CBFB = filter(offTarget, Gene == "CBFB")
# CBFB$num = 1:nrow(CBFB)
# 
# MYH = filter(offTarget, Gene == "MYH11")
# MYH$num = 1:nrow(MYH)
# 
# CBFBmap <- ggplot(data = CBFB) +
#   geom_point(aes(x = Chr_S, y = num, shape = factor(Date)), size = 3.5, color = "blue") +
#   geom_point(aes(x = Chr_E, y = num, shape = factor(Date)), size = 3.5, color = "red") +
#   xlim(c(rois$Start[rois$Gene == "CBFB"], rois$End[rois$Gene == "CBFB"])) +
#   geom_vline(aes(xintercept = 67111989), linetype = "dashed") +
#   geom_vline(aes(xintercept = 67111940), linetype = "dashed") +
#   ggtitle("CBFB off-target cuts") +
#   scale_shape_manual(values = c(0:2, 4:6, 15)) +
#   xlab("Chr coordinates")
# 
# MYHmap <- ggplot(data = MYH) +
#   geom_point(aes(x = Chr_S, y = num, shape = factor(Date)), size = 3.5, color = "blue") +
#   geom_point(aes(x = Chr_E, y = num, shape = factor(Date)), size = 3.5, color = "red") +
#   xlim(c(rois$Start[rois$Gene == "MYH11"], rois$End[rois$Gene == "MYH11"])) +
#   geom_vline(aes(xintercept = 15808795), linetype = "dashed") +
#   geom_vline(aes(xintercept = 15814171), linetype = "dashed") +
#   geom_vline(aes(xintercept = 15817006), linetype = "dashed") +
#   geom_vline(aes(xintercept = 15818634), linetype = "dashed") +
#   ggtitle("MYH off-target cuts") +
#   scale_shape_manual(values = c(0:2, 4:6, 15)) +
#   xlab("Chr coordinates")
# 
# 
# PMLmap <- ggplot(data = PML) +
#   geom_point(aes(x = Chr_S, y = num, shape = factor(Date)), size = 3.5, color = "blue") +
#   geom_point(aes(x = Chr_E, y = num, shape = factor(Date)), size = 3.5, color = "red") +
#   xlim(c(rois$Start[rois$Gene == "PML"], rois$End[rois$Gene == "PML"])) +
#   geom_vline(aes(xintercept = 74322670), linetype = "dashed") +
#   geom_vline(aes(xintercept = 74305671), linetype = "dashed") +
#   geom_vline(aes(xintercept = 74324389), linetype = "dashed") +
#   geom_vline(aes(xintercept = 74304134), linetype = "dashed") +
#   ggtitle("PML off-target cuts") +
#   scale_shape_manual(values = c(0:2, 4:6, 15)) +
#   xlab("Chr coordinates")
# 
# RARAmap <- ggplot(data = RARA) +
#   geom_point(aes(x = Chr_S, y = num, shape = factor(Date)), size = 3.5, color = "blue") +
#   geom_point(aes(x = Chr_E, y = num, shape = factor(Date)), size = 3.5, color = "red") +
#   xlim(c(rois$Start[rois$Gene == "RARA"], rois$End[rois$Gene == "RARA"])) +
#   geom_vline(aes(xintercept = 38505209), linetype = "dashed") +
#   geom_vline(aes(xintercept = 38511865), linetype = "dashed") +
#   ggtitle("RARA off-target cuts") +
#   scale_shape_manual(values = c(0:2, 4:6, 15)) +
#   xlab("Chr coordinates")
# 
# 
# export <- arrangeGrob(CBFBmap, MYHmap, PMLmap, RARAmap, nrow = 2, ncol = 2)
# 
# ggsave("offTargets-1.pdf", export, width = 11, height = 10)

# Karyogram
lims <- data.frame(paste0("chr", c(1:22, "X", "Y", "M")),
                   chrlims, c(1:22, "X", "Y", "M"), stringsAsFactors = F)
colnames(lims) <- c("Chr", "Size", "num")
lims$Chr <- factor(lims$Chr, levels = paste0("chr", c(1:22, "X", "Y", "M")))
lims$num <- factor(lims$num, levels = c(1:22))


test <- filter(allData, Date == 20190823)
test$ChrNum <- substring(test$Chr, 4)

test <- filter(test, ChrNum %in% c(1:22))

# This is gonna be a problem later because there are wayyy too many reads per experiment.
#   Figure out some sort of heatmap scheme...?
#   Also currently doesn't handle X, Y, M well
ggplot() +
  geom_segment(data = lims, aes(x = num, xend = num, y = 0, yend = Size),
               color = "lightgray", size = 3) +
  geom_segment(data = test, aes(x = as.integer(ChrNum) - 0.1,
                                xend = as.integer(ChrNum) + 0.1,
                                y = Chr_S, yend = Chr_S), color = "black") +
  coord_flip()

```

