---
title: "NanoporeCrunch"
output: html_document
---

```{r Libraries}
library(tidyverse)
library(gridExtra)
library(reshape2)
library(scales)
library(rlist)
library(readxl)
```


```{r Load csv}
# Load data, extract metadata from names
datasets <- list.files(path = ".", pattern = ".csv")
splits <- unlist(strsplit(datasets, "_"))

# Load supplementary files
expList <- read_excel("Experiment_List.xlsx")

  # Uses hg19  Not all cut sites matter for all runs
cutSites <- read_xlsx("Guides.xlsx")
cutSites <- data.frame(cutSites[,1:3], stringsAsFactors = F)
rownames(cutSites) <- NULL
cutSites$hg19 <- as.numeric(cutSites$hg19)
window = 50

rois <- read_xlsx("Regions_of_Interest.xlsx")


# Assign the easy things
dates <- grep("^[0-9]+$", splits, value = TRUE)
```

```{r Load experiment list}

# Next up: read these csvs and append metadata as new column
allData <- do.call(rbind, lapply(1:length(datasets), function(x) {
      working <- data.frame(read.csv(datasets[x], sep = ",", header = F, stringsAsFactors = F),
                            stringsAsFactors = F)
      working$Date <- dates[x]
      return(working)
      }
    ))

allData <- merge(allData, expList[, 1:3], by = "Date")

colnames(allData)[1:7] <- c("Date", "Chr", "Chr_S", "Chr_E", "Name", "Score", "Strand")

# Remove weird special characters that are somehow getting prepended onto each data set
allData$Chr <- gsub("[^a-zA-Z0-9]", "", allData$Chr)

# Chromosome lengths for plotting purposes
chrlims <- c(247250000, 242952000, 199502000, 191274000, 180858000,
             170900000, 158822000, 146275000, 140274000, 135375000,
             134453000, 132350000, 114143000, 106369000, 100339000,
             88828000, 78775000, 76118000, 63812000, 62436000,
             46945000, 49692000, 154914000, 57773000)
# Last two X, Y.  Need to add M at some point


# Length will be important later
  # Semi-arbitrarily filter at 50
allData <- allData %>% mutate(Len = Chr_E - Chr_S) %>% filter(Score >= 50)
```


```{r Get dataframe of bridging reads (same read name, both 9 and 22) }
# Seems like names are unique.  Go off that assumption for now

chr9reads <- allData %>% filter(Chr == "chr9")
chr9reads <- unique(chr9reads$Name)

chr13reads <- allData %>% filter(Chr == "chr13")
chr13reads <- unique(chr13reads$Name)

chr15reads <- allData %>% filter(Chr == "chr15")
chr15reads <- unique(chr15reads$Name)

chr16reads <- allData %>% filter(Chr == "chr16")
chr16reads <- unique(chr16reads$Name)

chr17reads <- allData %>% filter(Chr == "chr17")
chr17reads <- unique(chr17reads$Name)

chr19reads <- allData %>% filter(Chr == "chr19")
chr19reads <- unique(chr19reads$Name)

chr22reads <- allData %>% filter(Chr == "chr22")
chr22reads <- unique(chr22reads$Name)

chr922int <- intersect(chr9reads, chr22reads)
chr1517int <- intersect(chr15reads, chr17reads)

dataUnion <- allData %>% filter(Chr %in% c("chr9", "chr13", "chr15", "chr16",
                                             "chr17", "chr19", "chr22"))

```


```{r Crunch it!}

# Create columns for *all* cut sites
    # Later we can prune by which guides were present in which experiment

# Search Chr_S and Chr_E for proximity to each cut site

# Problem: cut sites may match locations on other chromosomes

# For each guide
  # Filter dataUnion by Chr
  # Filter to dates in which that guide was used


cutDists <- lapply(1:nrow(cutSites), function(x) {
  working <- allData %>% filter(Chr == cutSites$Chr[x])
  dateList <- expList$Date[!(is.na(expList[, cutSites$Name[x]]))]
  working2 <- filter(working, as.numeric(Date) %in% dateList)
  working2[, paste0(cutSites$Name[x],"S")] <- c(abs(working2$Chr_S - cutSites$hg19[x]))
  working2[, paste0(cutSites$Name[x],"E")] <- c(abs(working2$Chr_E - cutSites$hg19[x]))
  working2
  }
)

allDists <- join_all(cutDists, type = "full")

allDists$rowMin <- sapply(1:nrow(allDists), function(x)
              min(allDists[x, 11:ncol(allDists)], na.rm = T))

# Filter by distance from cut site

withCuts <- filter(allDists, rowMin < 50)

# Whew.  Works up to here

withCuts$cutGuide <- sapply(1:nrow(withCuts), function(x)
        colnames(withCuts[x, 11:76])[match(withCuts$rowMin[x], withCuts[x, 11:76])])


whichCuts <- withCuts %>% select(c("Date", "Chr", "Chr_S", "Chr_E", "Name", "Score", "Strand",
                                   "Sample", "Device", "Len", "cutGuide")) %>% filter(Score >= 50)


# Collapse by start or end first

test <- whichCuts
test <- test %>% mutate(End = substr(cutGuide, nchar(cutGuide), nchar(cutGuide))) %>%
  mutate(cutGuide = substr(cutGuide, 0, nchar(cutGuide)-1))


test2 <- test %>% group_by(Date, cutGuide, Device) %>% dplyr::summarise(n())
test3 <- test %>% group_by(Date, Device) %>% dplyr::summarise(n())


# numCuts <- rightCuts %>% group_by(Date, Guide) %>% summarise(Num = n())
# 
# nReads <- allData %>% group_by(Date) %>% summarise(totReads = n())
# 
# numCuts <- numCuts %>% group_by(Date) %>%
#   mutate(nReads = nReads$totReads[match(Date, nReads$Date)]) %>%
#   mutate(scaledReads = Num/sum(unique(nReads)))
# 
# ggplot(numCuts) + geom_point(aes(x = Guide, y = scaledReads, color = factor(Date)))

```


```{r Check cuts from other guides on regions of interest}
# There's probably a way of doing all this stuff with `join`
  # See near the bottom of this chunk...

# Region of interest dataframe, then iterate over guides to see when they're cutting there

inGenes <- bind_rows(lapply(1:nrow(rois), function(x) {
  
            working <- allData %>% filter(Chr == rois$Chr[x])
              
              # Could also be intersection of:
                # Rows of rois where Chr_S >= start
                # Rows of rois where Chr_E <= end
              # and other two possibilities

            test <- intersect(which(working$Chr_S >= rois$Start[x]),
                            which(working$Chr_S <= rois$End[x]))
            test2 <- intersect(which(working$Chr_E >= rois$Start[x]),
                            which(working$Chr_E <= rois$End[x]))
            test3 <- intersect(which(working$Chr_S <= rois$Start[x]),
                            which(working$Chr_E >= rois$End[x]))
            
            return(working[unique(c(test, test2, test3)), ])
              # 
              # 
              # mutate(inRoi = ifelse((Chr_S >= rois$Start[x] & Chr_S <= rois$End[x]) |
              #                         (Chr_E >= rois$Start[x] & Chr_E <= rois$End[x]) |
              #                         (Chr_S <= rois$Start[x] & Chr_E >= rois$End[x]), 1, 0))
      }
  ))

# working <- distinct(testing)

# Match inRoi to genes

  # Look for chromosomes, 

inGenes$endint <- sapply(1:nrow(inGenes), function(x) {
  endint = intersect(which(inGenes$Chr_E[x] >= rois$Start), which(inGenes$Chr_E[x] <= rois$End))
  endint = ifelse(inGenes$Chr[x] == rois$Chr[endint], endint, 0)
  endint = ifelse(length(endint) > 1 && 0 %in% endint, sum(endint), endint)
})

inGenes$startint <- sapply(1:nrow(inGenes), function(x) {
  startint = intersect(which(inGenes$Chr_S[x] >= rois$Start), which(inGenes$Chr_S[x] <= rois$End))
  startint = ifelse(inGenes$Chr[x] == rois$Chr[startint], startint, 0)
  startint = ifelse(length(startint) > 1 && 0 %in% startint, sum(startint), startint)
})

inGenes$both <- sapply(1:nrow(inGenes), function(x) {
  both = intersect(which(inGenes$Chr_S[x] <= rois$Start), which(inGenes$Chr_E[x] >= rois$End))
  both = ifelse(inGenes$Chr[x] == rois$Chr[both], both, 0)
  both = ifelse(length(both) > 1 && 0 %in% both, sum(both), both)
})

inGenes <- inGenes %>% mutate(Gene = pmax(startint, endint, both, na.rm = T)) %>%
  mutate(Gene = rois$Gene[Gene]) %>% select(-startint, -endint, -both, -inRoi)


expGenes <- c(sapply(1:nrow(expList), function(x) {
  test <- colnames(expList)[!is.na(expList[x, ])]
  test <- test[5:length(test)]
  test2 <- substr(test, 0, 3)
  rois$Gene[match(unique(test2),substr(rois$Gene, 0, 3))]
}))

expGenes <- data.frame(cbind(unlist(expGenes),
                             rep(expList$Date, times = lengths(expGenes))),
                       stringsAsFactors = F)
colnames(expGenes) <- c("Gene", "Date")

test2 <- inner_join(inGenes, expGenes) # Gene/Date combos for which there are reads

# Works until here

test3 <- test2 %>% group_by(Date) %>% dplyr::summarise(n())
test4 <- allData %>% group_by(Date) %>% dplyr::summarise(reads = n(), totLen = sum(Len))
test3$reads <- filter(test4, Date %in% test3$Date)$reads

ggplot(test3) + geom_point(aes(x = Date, y = `n()`/reads)) + ylim(c(0, 0.004))

working <- allData %>% filter(Date == 20190501 & Chr == "chr21")

ggplot(working) + geom_histogram(aes(x = Chr_S), bins = 100) + xlim(c(36160098, 37376965))

# By each date, remove 

# working <- rois %>% 
#   group_by(r=row_number()) %>% 
#   mutate(custom = list(Start:End)) %>% 
#   ungroup %>% select(-r) %>% 
#   unnest()
# working <- working$custom
# 
# working2 <- allData %>% mutate(overlap = ifelse(Chr_S %in% working | Chr_E %in% working, 1, 0))
# working2 <- working2 %>% filter(overlap == 1)

```


```{r CEBPA cuts}
chr19Cuts <- chr19df %>% filter(Chr == "chr19") %>%
  mutate(CEBPA1 = ifelse((abs(33786401-Chr_E) < window) |
                          (abs(33786401-Chr_S) < window), 1, 0)) %>%
  mutate(CEBPA2 = ifelse((abs(33794671-Chr_E) < window) |
                          (abs(33794671-Chr_S) < window), 1, 0)) %>%
  mutate(anyCEBPA = ifelse(CEBPA1 | CEBPA2, 1, 0)) %>%
  filter(anyCEBPA == 1) %>% select(-Strand, -Score, -anyCEBPA)

cebpaCuts <- melt(chr19Cuts, id = c("Chr", "Chr_S", "Chr_E", "Name")) %>% filter(value == 1) %>%
  group_by(variable) %>% mutate(Len = Chr_E - Chr_S) %>% summarise(n())
```

```{r FLT3 cuts}
chr13Cuts <- chr13df %>% filter(Chr == "chr13") %>%
  mutate(FLT3F1 = ifelse((abs(28609169-Chr_E) < window) |
                          (abs(28609169-Chr_S) < window), 1, 0)) %>%
  mutate(FLT3R1 = ifelse((abs(28591129-Chr_E) < window) |
                          (abs(28591129-Chr_S) < window), 1, 0)) %>%
  mutate(FLT3R2 = ifelse((abs(28590244-Chr_E) < window) |
                          (abs(28590244-Chr_S) < window), 1, 0)) %>%
  mutate(anyFLT3 = ifelse(FLT3F1 | FLT3R1 | FLT3R2, 1, 0)) %>%
  filter(anyFLT3 == 1) %>% select(-Strand, -Score, -anyFLT3)

flt3Cuts <- melt(chr13Cuts, id = c("Chr", "Chr_S", "Chr_E", "Name")) %>% filter(value == 1) %>%
  group_by(variable) %>% mutate(Len = Chr_E - Chr_S) %>% summarise(n())

```


```{r Get length of reads with cuts}
# The next four bits are for the table Olga's using in the abstract


# Total number of cuts by guides
test9 <- df %>% filter(Chr == "chr9") %>%
  mutate(ABL11 = ifelse((abs(133613086-Chr_E) < window) |
                          (abs(133613086-Chr_S) < window), 1, 0)) %>%
  mutate(ABL12 = ifelse((abs(133655785-Chr_E) < window) |
                          (abs(133655785-Chr_S) < window), 1, 0)) %>%
  mutate(ABL13 = ifelse((abs(133691503-Chr_E) < window) |
                          (abs(133691503-Chr_S) < window), 1, 0)) %>%
  mutate(ABL14 = ifelse((abs(133730289-Chr_E) < window) |
                          (abs(133730289-Chr_S) < window), 1, 0)) %>%
  mutate(ABL15 = ifelse((abs(133732446-Chr_E) < window) |
                          (abs(133732446-Chr_S) < window), 1, 0)) %>%
  mutate(anyABL = ifelse(ABL11 | ABL12 | ABL13 | ABL14 | ABL15, 1, 0)) %>%
  filter(anyABL == 1) %>% select(-Strand, -Score, -anyABL)

test9Cuts <- melt(test9, id = c("Chr", "Chr_S", "Chr_E", "Name", "Len")) %>% filter(value != 0) %>%
  group_by(variable) %>% summarise(n())

test22 <- df %>% filter(Chr == "chr22") %>%
  mutate(BCR1 = ifelse((abs(23630389-Chr_E) < window) |
                         (abs(23630389-Chr_S) < window), 1, 0)) %>%
  mutate(BCR2 = ifelse((abs(23630060-Chr_E) < window) |
                         (abs(23630060-Chr_S) < window), 1, 0)) %>%
  mutate(anyBCR = ifelse(BCR1 | BCR2, 1, 0)) %>% filter(anyBCR == 1) %>% select(-Strand, -Score, -anyBCR)

test22Cuts <- melt(chr22Cuts, id = c("Chr", "Chr_S", "Chr_E", "Name", "Len")) %>% filter(value != 0) %>%
  group_by(variable) %>% summarise(n())

print(paste0("Cuts by guides: ", sum(test9Cuts$`n()`, test22Cuts$`n()`)))

# Reads on target with start at any cut site
# Rerun above bits, but without "summarise"

print(paste0("On-target, have cut: ", length(unique(c(test9Cuts$Name, test22Cuts$Name)))))

# Reads with chr9 and chr22

print(paste0("On both 9 and 22: ", length(unique(c(chr22Cuts$Name, chr9Cuts$Name)))))




# Other crap

chr9Cuts <- chr9Cuts %>% mutate(Len = Chr_E - Chr_S)

chr22Cuts <- chr22Cuts %>% mutate(Len = Chr_E - Chr_S)

bpOnTarget <- sum(chr9Cuts$Len) + sum(chr22Cuts$Len) + sum(chr13Cuts$Len) + sum(chr19Cuts$Len)

print(paste0("Base Pairs on Target: ", bpOnTarget))



# KCL22 and KU812
# allCuts9 <- chr9Cuts %>% select(-ABL11, -ABL12, -ABL13, -ABL14, -ABL15)
# allCuts22 <- chr22Cuts %>% select(-BCR1, -BCR2)

# K562
allCuts9 <- chr9Cuts %>% select(-ABL11, -ABL12, -ABL13, -ABL14, -ABL15)
allCuts22 <- chr22Cuts %>% select(-BCR1, -BCR2)




allCuts <- rbind(allCuts9, allCuts22)

allCuts <- spread(allCuts, key = "Chr", value = "Len") %>%
      mutate(TotalLen = ifelse(!is.na(chr22), chr22, 0)+ifelse(!is.na(chr9), chr9, 0))


# spreadBridging <- bridging %>% filter(Score > 50) %>% select(-Chr_S, -Chr_E, -Score, -Strand)

# spreadBridging <- spreadBridging[c(-157, -266), ]  # Weird ones for KCL22 run
# spreadBridging <- spreadBridging[-55, ] # Weird one for KU812 run
# spreadBridging <- spreadBridging[c(-16, -11, -13, -14, -23), ]  # Weird ones for K562 run

# spreadBridging <- spread(spreadBridging, key = "Chr", value = "Len") %>%
#       mutate(TotalLen = ifelse(!is.na(chr22), chr22, 0)+ifelse(!is.na(chr9), chr9, 0))

bridging <- allCuts %>% group_by(Name) %>% summarise(num = n()) %>% filter(num == 2)
bridging <- bridging$Name

bridging <- allCuts %>% filter(Name %in% bridging) %>% group_by(Name)

ggplot(subset(bridging, !is.na(chr22))) +
  geom_histogram(aes(x = Chr_S), fill = "green", alpha = 0.5, bins = 100) +
  geom_histogram(aes(x = Chr_E), fill = "red", alpha = 0.5, bins = 100)

ggplot(subset(bridging, !is.na(chr9))) +
  geom_histogram(aes(x = Chr_S), fill = "green", alpha = 0.5, bins = 100) +
  geom_histogram(aes(x = Chr_E), fill = "red", alpha = 0.5, bins = 100)
```




```{r Save a CSV of cuts per guide}
guideTable <- rbind(ablCuts, bcrCuts)
colnames(guideTable) <- c("Guide", "nCuts")
write.csv(guideTable, paste0("guideTable_", refmat,".csv"))
```

